File: server/game-server/app/game/gameController.ts
import Battle from "./battle";
import { IPlayerInfo } from "./battlePlayer";
export interface IUserCommand {
    uid: string;
    battleId: string;
    commandId: COMMAND_ID;
    args?: any;
}
export enum COMMAND_ID {
    USE_HAND_CARD,
}
export default class GameController {
    public static getInstance() {
        if (!this._instance) { this._instance = new this(); }
        return this._instance;
    }
    private static _instance: GameController;
    private battleMap: { [battleId: string]: Battle };
    private constructor() {
        this.battleMap = {};
    }
    public createBattle(seed: number) {
        const battle = new Battle(seed);
        this.battleMap[battle.id] = battle;
        return battle.id;
    }
    public startBattle(battleId: string) {
        const b = this.getBattle(battleId);
        const result = b.start();
        return result;
    }
    public setPlayer(battleId: string, palyerInfos: IPlayerInfo[]) {
        const b = this.getBattle(battleId);
        b.setPlayer(palyerInfos);
        console.log(battleId, "setInfos!");
    }
    public command(command: IUserCommand) {
        const battle = this.getBattle(command.battleId);
        if (!battle) { return; }
        const player = battle.getPlayer(command.uid);
        if (!player) { return; }
        battle.command(command);
    }
    private getBattle(battleId: string) {
        return this.battleMap[battleId];
    }
}

File: server/Logic/Game/GameController.cs
using System.Collections;
using System.Collections.Generic;
namespace Logic
{

	class GameController
	{
		private static GameController _instance;
		private Dictionary<string, Battle> battleMap;
		private GameController()
		{
			this.battleMap = new Dictionary<string, Battle>();
		}
		public static GameController getInstance()
		{
			if (_instance == null) { _instance = new GameController(); }
			return _instance;
		}
		public string createBattle(int seed)
		{
			Battle battle = new Battle(seed);
			this.battleMap.TryAdd(battle.id, battle);
			return battle.id;
		}
		public void startBattle(string battleId)
		{
			Battle b = this.getBattle(battleId);
			b.start();
		}
		public void setPlayer(string battleId, PlayerInfo[] playerInfos)
		{
			Battle b = this.getBattle(battleId);
			b.setPlayers(playerInfos);
			System.Console.WriteLine("Battle:{0} setPlayerInfos!", battleId);
		}

		public void command(UserCommand command)
		{
			Battle battle = this.getBattle(command.battleId);
			if (battle == null) { return; }
			BattlePlayer player = battle.getPlayer(command.uid);
			if (player == null) { return; }
			battle.command(command);
		}
		public void getSituation(string battleId)
		{

		}

		public Battle getBattle(string battleId)
		{
			Battle ret;
			this.battleMap.TryGetValue(battleId, out ret);
			return ret;
		}
	}

}